### ISSUE TITLE
Listener IdleTimeout

### ISSUE THREAD

**Issue by stevesloka** [heptio/contour#1054]
**Describe the solution you'd like**
Browsers such as Chrome have a tendency to keep TCP connections open after the HTTP request-response cycle is done to be reused by future requests and improve the UX. Since Envoy's default is *no timeout*, connections can stay open indefinitely. In high-traffic environments, this can cause Envoy's memory consumption to grow very quickly and become unmanageable.

Reported by @rata & Johanan Liebermann in #contour slack (https://kubernetes.slack.com/archives/C8XRH2R4J/p1556480572023000), they provided a diff of what change they made to validate the issue:

```
diff --git a/internal/envoy/listener.go b/internal/envoy/listener.go
index 9001322..97bd94e 100644
--- a/internal/envoy/listener.go
+++ b/internal/envoy/listener.go
@@ -103,6 +103,7 @@ func HTTPConnectionManager(routename, accessLogPath string) listener.Filter {
                                        "access_log":         accesslog(accessLogPath),
                                        "use_remote_address": {Kind: &types.Value_BoolValue{BoolValue: true}}, // TODO(jbeda) should this ever be false?
                                        "normalize_path":     {Kind: &types.Value_BoolValue{BoolValue: true}},
+                                       "idle_timeout":       sv("60s"),
                                },
                        },
                },
```

**Anything else you would like to add:**
We should allow this value to be customized via an arg that gets passed to Contour. Also, this should be applied to all listeners that get created (i.e. http/https). It might be good to come up with a sane default that would apply to a majority of environments.

**Comment by mattalberts** [comment]
This feels similar to #944, but I leaned on the pre_route variant to give it a more granular effect.
 - https://github.com/mattalberts/contour/commit/aca3cb958959702fe43af4d40d074b5c22384970

On top of that, I layered on the listener `request_timeout` to give a global control:
 - https://github.com/mattalberts/contour/commit/6123f5ddd5408d57f4a16f4851672668b2f8910b

I can bring these over to master (I patched them on-top of 0.10.0)?

**Comment by stevesloka** [comment]
Ahh cool, not sure how I missed that other issue @mattalberts. I'm curious what @davecheney thinks, but having a default overall setting and allowing each route to override sounds like a good impl to me.

**Comment by johananl** [comment]
@stevesloka @mattalberts I would like to emphasize I was talking about the **listener-level** [idle_timeout](https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/network/http_connection_manager/v2/http_connection_manager.proto.html?#envoy-api-field-config-filter-network-http-connection-manager-v2-httpconnectionmanager-idle-timeout) Envoy option, which seems to be different than the **route-level** [idle_timeout](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route.proto#envoy-api-field-route-routeaction-idle-timeout).

I think these two options are rather confusing and don't do the same thing.

From the docs:

## Listener-level `idle_timeout`

>The idle timeout for connections managed by the connection manager. The idle timeout is defined as **the period in which there are no active requests**. If not set, there is no idle timeout. When the idle timeout is reached **the connection will be closed**. If the connection is an HTTP/2 connection a drain sequence will occur prior to closing the connection. See drain_timeout.

## Route-level `idle_timeout`

>The idle timeout is distinct to timeout, which provides an upper bound on the upstream response time; idle_timeout instead bounds the amount of time the **request’s stream** may be idle.

>After header decoding, the idle timeout will apply on downstream and upstream request events. Each time an encode/decode event for headers or data is processed for the stream, the timer will be reset. If the timeout fires, the stream is **terminated with a 408 Request Timeout error code** if no upstream response header has been received, otherwise a stream reset occurs.

It's the listener-level idle timeout which causes connections to accumulate in Envoys memory, not the route-level ones. I was able to verify that setting the route-level idle timeout doesn't affect the memory consumption problem we've observed, whereas setting the listener-level one immediately fixes the problem.

## Reproduction Case

You can use the following reproduction case to verify Envoy's behavior with the two options as well as its default behavior.

### No idle timeout

```bash
cat <<EOF >envoy.yaml
admin:
  access_log_path: /tmp/admin_access.log
  address:
    socket_address:
      protocol: TCP
      address: 0.0.0.0
      port_value: 9901
static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address:
        protocol: TCP
        address: 0.0.0.0
        port_value: 10000
    filter_chains:
    - filters:
      - name: envoy.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager
          stat_prefix: ingress_http
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              - match:
                  prefix: "/"
                route:
                  host_rewrite: www.google.com
                  cluster: service_google
          http_filters:
          - name: envoy.router
  clusters:
  - name: service_google
    connect_timeout: 0.25s
    type: LOGICAL_DNS
    # Comment out the following line to test on v6 networks
    dns_lookup_family: V4_ONLY
    lb_policy: ROUND_ROBIN
    load_assignment:
      cluster_name: service_google
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: www.google.com
                port_value: 443
    tls_context:
      sni: www.google.com
EOF

docker run --rm -d -p 10000:10000 -p 9901:9901 -v $(pwd)/envoy.yaml:/etc/envoy/envoy.yaml --name envoy envoyproxy/envoy-dev:b8325ac4167284b4d566c2779c15ec3c8473dc66
```

**Browse http://localhost:10000 using Chromium for example**. `curl` doesn't keep the connection open after the HTTP request is done.

```
for i in {1..100}; do date; curl -s localhost:9901/stats | grep total_conn; sleep 5; done
Tue 30 Apr 2019 10:58:59 PM CEST
server.total_connections: 2
Tue 30 Apr 2019 10:59:04 PM CEST
server.total_connections: 2
Tue 30 Apr 2019 10:59:09 PM CEST
server.total_connections: 2
Tue 30 Apr 2019 10:59:14 PM CEST
server.total_connections: 2
Tue 30 Apr 2019 10:59:19 PM CEST
server.total_connections: 2
Tue 30 Apr 2019 10:59:24 PM CEST
server.total_connections: 2
Tue 30 Apr 2019 10:59:29 PM CEST
server.total_connections: 2
```

### Listener-level idle timeout

```bash
cat <<EOF >envoy.yaml
admin:
  access_log_path: /tmp/admin_access.log
  address:
    socket_address:
      protocol: TCP
      address: 0.0.0.0
      port_value: 9901
static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address:
        protocol: TCP
        address: 0.0.0.0
        port_value: 10000
    filter_chains:
    - filters:
      - name: envoy.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager
          stat_prefix: ingress_http
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              - match:
                  prefix: "/"
                route:
                  host_rewrite: www.google.com
                  cluster: service_google
          http_filters:
          - name: envoy.router
          idle_timeout: 5s
  clusters:
  - name: service_google
    connect_timeout: 0.25s
    type: LOGICAL_DNS
    # Comment out the following line to test on v6 networks
    dns_lookup_family: V4_ONLY
    lb_policy: ROUND_ROBIN
    load_assignment:
      cluster_name: service_google
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: www.google.com
                port_value: 443
    tls_context:
      sni: www.google.com
EOF

docker run --rm -d -p 10000:10000 -p 9901:9901 -v $(pwd)/envoy.yaml:/etc/envoy/envoy.yaml --name envoy envoyproxy/envoy-dev:b8325ac4167284b4d566c2779c15ec3c8473dc66
```

**Browse http://localhost:10000 using Chromium for example**. `curl` doesn't keep the connection open after the HTTP request is done.

```
for i in {1..100}; do date; curl -s localhost:9901/stats | grep total_conn; sleep 5; done
Tue 30 Apr 2019 11:00:46 PM CEST
server.total_connections: 2
Tue 30 Apr 2019 11:00:51 PM CEST
server.total_connections: 2
Tue 30 Apr 2019 11:00:56 PM CEST
server.total_connections: 2
Tue 30 Apr 2019 11:01:01 PM CEST
server.total_connections: 0
Tue 30 Apr 2019 11:01:06 PM CEST
server.total_connections: 0
```

### Route-level idle timeout

```bash
cat <<EOF >envoy.yaml
admin:
  access_log_path: /tmp/admin_access.log
  address:
    socket_address:
      protocol: TCP
      address: 0.0.0.0
      port_value: 9901
static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address:
        protocol: TCP
        address: 0.0.0.0
        port_value: 10000
    filter_chains:
    - filters:
      - name: envoy.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager
          stat_prefix: ingress_http
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              - match:
                  prefix: "/"
                route:
                  host_rewrite: www.google.com
                  cluster: service_google
                  idle_timeout: 5s
          http_filters:
          - name: envoy.router
  clusters:
  - name: service_google
    connect_timeout: 0.25s
    type: LOGICAL_DNS
    # Comment out the following line to test on v6 networks
    dns_lookup_family: V4_ONLY
    lb_policy: ROUND_ROBIN
    load_assignment:
      cluster_name: service_google
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: www.google.com
                port_value: 443
    tls_context:
      sni: www.google.com
EOF

docker run --rm -d -p 10000:10000 -p 9901:9901 -v $(pwd)/envoy.yaml:/etc/envoy/envoy.yaml --name envoy envoyproxy/envoy-dev:b8325ac4167284b4d566c2779c15ec3c8473dc66
```

**Browse http://localhost:10000 using Chromium for example**. `curl` doesn't keep the connection open after the HTTP request is done.

```
for i in {1..100}; do date; curl -s localhost:9901/stats | grep total_conn; sleep 5; done
Tue 30 Apr 2019 11:08:05 PM CEST
server.total_connections: 3
Tue 30 Apr 2019 11:08:10 PM CEST
server.total_connections: 3
Tue 30 Apr 2019 11:08:15 PM CEST
server.total_connections: 3
Tue 30 Apr 2019 11:08:20 PM CEST
server.total_connections: 3
Tue 30 Apr 2019 11:08:25 PM CEST
server.total_connections: 3
Tue 30 Apr 2019 11:08:30 PM CEST
server.total_connections: 3
Tue 30 Apr 2019 11:08:35 PM CEST
server.total_connections: 3
Tue 30 Apr 2019 11:08:40 PM CEST
server.total_connections: 3
Tue 30 Apr 2019 11:08:45 PM CEST
server.total_connections: 3
```

>I'm curious what @davecheney thinks, but having a default overall setting and allowing each route to override sounds like a good impl to me.

IMHO a global Contour default with an option to override at the Ingress/IngressRoute level sounds good.

**Comment by mattalberts** [comment]
Understood, my plan was to put together three PRs,
 - Listener `idle_timeout`
 - Listener `request_timeout`
 - Route `idle_timeout`

I just wanted to wait on @davecheney's input around the implementation.  The PRs are easy :)

**Comment by davecheney** [comment]
@mattalberts please set 

- listener idle_timeout to 60 seconds. Please do not make this configurable
- listener request_timeout to 15 seconds. Please do not make this configurable.

Please move the discussion of route idle_timeout to another issue, possibly #1052
